### ISSUE TITLE
Should we flush memtable(s) when cleaning up a column family?

### ISSUE THREAD

**Issue by raphaelsc** [scylladb/scylladb#1239]
It was suggested that memtable should be flushed in event of a 'nodetool cleanup' Why? Keys in memtable may no longer belong to node. I took a look at C*'s code, and AFAICT, they ignore this particular issue. Their implementation of cleanup can be found at: src/java/org/apache/cassandra/db/compaction/CompactionManager.java

**Comment by slivne** [comment]
I think its hard to understand if this is an issue or not without a sample that shows this is an issue.

I think the following may be an issue

1. We start with a single node write A,B,C
2. We add a node and B,C transfer to node 2
3. on node 2 we delete B - and flush memtables.
4. gc_period passes and B is removed completely
5. node B is decomissioned
6. node 1 received ownership of C - yet B is resurected 


We had a similar issue with caches and we have verified that when ownership is returned (e.g. node2 is decommissioned in our scenario) the cache is cleared. 

If B is resurrected - then there is an issue - in C* I think streaming requires files - so flush may be done automatically when a node is added and after transfer of token ownership is complete  - maybe we need todo the same. The issue may be broader then cleanup flushing/not flushing.

**Comment by slivne** [comment]
need to check what we do in the above scenario to decide the priority of this issue

**Comment by tzach** [comment]
@raphaelsc what is the affect of keeping a no longer relevant data in the memtable?

**Comment by dorlaor** [comment]
?

**Comment by tgrabiec** [comment]
@tzach The effect could be that queries may start to return data which was deleted. See Shlomi's example.

**Comment by bhalevy** [comment]
But can't we return those token from the memtable before it's flushed?
Should we filter those tokens on the read path when reading from either the memtable or even from a sstable?
Can we filter them out when flushing the memtable regardless of `nodetool cleanup`?

**Comment by tgrabiec** [comment]
Filtering is not necessary, the coordinator is not supposed to request data from the replica which is not owned by it. Cleanup drops keys which are no longer owned by the replica, so no read should touch them. Replica can re-obtain ownership only after cleanup is finished, we prohibit topology changes concurrent with cleanup.

**Comment by bhalevy** [comment]
> Filtering is not necessary, the coordinator is not supposed to request data from the replica which is not owned by it. Cleanup drops keys which are no longer owned by the replica, so no read should touch them. Replica can re-obtain ownership only after cleanup is finished, we prohibit topology changes concurrent with cleanup.

In this case I agree we need to flush before cleanup

**Comment by avikivity** [comment]
@bhalevy should we backport this? I think we should, it's a correctness issue, even if it's rare and not a regression.

**Comment by bhalevy** [comment]
> @bhalevy should we backport this? I think we should, it's a correctness issue, even if it's rare and not a regression.

Yes. I concur.

**Comment by avikivity** [comment]
Unfortunately, it doesn't backport cleanly due to fc278be6c4edc028c7908d3c2d236b479032b3f8

**Comment by bhalevy** [comment]
> Unfortunately, it doesn't backport cleanly due to [fc278be](https://github.com/scylladb/scylladb/commit/fc278be6c4edc028c7908d3c2d236b479032b3f8)

I'll prepare a backport branch then

**Comment by denesb** [comment]
> @denesb was this backported to 5.0 but not 5.1?

It looks like so. I assumed it is already in 5.1, I should have checked.
@bhalevy is https://github.com/scylladb/scylladb/pull/12415 also good for 5.1? If now please prepare another backport PR.

**Comment by bhalevy** [comment]
No, I'll prepare a backport branch
